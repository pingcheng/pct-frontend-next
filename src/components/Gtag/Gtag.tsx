import Script from "next/script";

export default function Gtag() {
  const gaMeasurementId = process.env.NEXT_PUBLIC_GA_MEASUREMENT_ID;

  if (!gaMeasurementId) {
    return <></>;
  }

  return (
    <>
      <Script
        src={`https://www.googletagmanager.com/gtag/js?id=${gaMeasurementId}`}
      />
      <Script id="google-anlytics">
        {`
          window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());

          gtag('config', '${gaMeasurementId}');
        `}
      </Script>
    </>
  );
}
