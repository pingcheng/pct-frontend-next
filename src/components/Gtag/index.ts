import Gtag from "./Gtag";

export default Gtag;
