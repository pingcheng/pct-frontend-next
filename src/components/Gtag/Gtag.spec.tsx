import { render } from "@testing-library/react";
import Gtag from "./index";
import Script from "next/script";

// Mock the Next.js Script component
jest.mock("next/script", () => {
  return jest.fn().mockImplementation(({ children, src, id }) => {
    return children ? (
      <script id={id} dangerouslySetInnerHTML={{ __html: children }} />
    ) : (
      <script defer src={src} />
    );
  });
});

describe("test <Gtag />", () => {
  const originalEnv = process.env;

  beforeEach(() => {
    jest.resetAllMocks();
    // Clear the env variable before each test
    delete process.env.NEXT_PUBLIC_GA_MEASUREMENT_ID;
  });

  afterAll(() => {
    // Restore the env
    process.env = originalEnv;
  });

  test("should render nothing when GA_MEASUREMENT_ID is not set", () => {
    const { container } = render(<Gtag />);
    expect(container.firstChild).toBeNull();
    expect(Script).not.toHaveBeenCalled();
  });

  test("should render GA scripts when GA_MEASUREMENT_ID is set", () => {
    // Set the env variable
    process.env.NEXT_PUBLIC_GA_MEASUREMENT_ID = "G-12345";

    const { container } = render(<Gtag />);

    // Check if Script component was called twice (for the two scripts)
    expect(Script).toHaveBeenCalledTimes(2);

    // Snapshot test to ensure consistent rendering
    expect(container).toMatchSnapshot();
  });
});
