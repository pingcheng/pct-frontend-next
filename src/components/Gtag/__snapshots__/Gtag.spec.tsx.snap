// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`test <Gtag /> should render GA scripts when GA_MEASUREMENT_ID is set 1`] = `<div />`;
